src/d3a/models/strategy/external_strategies/__init__.py

```diff
@@ -444,38 +442,101 @@ def _set_energy_forecast(self, payload):
                 IncomingRequest("set_energy_forecast", arguments,
                                 energy_forecast_response_channel))
 
-    def _set_energy_forecast_impl(self, arguments, response_channel):
+    def _set_energy_measurement(self, payload: Dict) -> None:
+        """Callback of set_energy_measurement redis endpoint."""
+        transaction_id = self._get_transaction_id(payload)
+        energy_measurement_response_channel = \
+            f"{self.channel_prefix}/response/set_energy_measurement"
+        try:
+            arguments = json.loads(payload["data"])
+            assert set(arguments.keys()) == {"energy_measurement", "transaction_id"}
+        except Exception:
+            logging.exception(
+                f"Incorrect _set_energy_measurement request. Payload: {payload}.")
+            self.redis.publish_json(
+                energy_measurement_response_channel,
+                {"command": "set_energy_measurement",
+                 "error": "Incorrect _set_energy_measurement request. "
+                          "Available parameters: (energy_measurement).",
+                 "transaction_id": transaction_id})
+        else:
+            self.pending_requests.append(
+                IncomingRequest("set_energy_measurement", arguments,
+                                energy_measurement_response_channel))
+
+    @staticmethod
+    def _validate_values_positive_in_profile(profile: Dict) -> None:
+        """Validate whether all values are positive in a profile."""
+        for time_str, energy in profile.items():
+            assert energy >= 0.0, f"Energy is not positive for time stamp {time_str}."
+
+    def _set_energy_forecast_impl(self, arguments: Dict, response_channel: str) -> None:
+        """
+        Digest command from buffer and perform set_energy_forecast.
+        Args:
+            arguments: Dictionary containing "energy_forecast" that is a dict containing a profile
+                key: time string, value: energy in kWh
+            response_channel: redis channel string where the response should be sent to
+        """
         try:
-            assert arguments["energy_forecast"] >= 0.0
-            self.energy_forecast_buffer_Wh = arguments["energy_forecast"]
+            self._validate_values_positive_in_profile(arguments["energy_forecast"])
```

Comment:
I have the feeling that this won't work for Home Meter devices (since the profile values can be both positive and negative).

Reply:
Indeed it wont work. We dont have a external HomeMeter yet. Shall we change it in the frame of https://gridsingularity.atlassian.net/browse/D3ASIM-3525 ?

Reply:
That's an option. The reason why I pointed this out here is that the `ExternalMixin` is defining methods that are incompatible (or not used) with the classes that inherit from it. That's what I also meant in the comment above: https://github.com/gridsingularity/d3a/pull/1200#discussion_r690140977

Reply:
Ok, let me try something.

Reply:
Hmmm, I tried, but I can only think of a way that will lead to subclass explosion (that we discussed in the workshop). This is mainly due to the fact that we have this special architecture of subclassing in the external strategies.
Injecting the measurement and forecast endpoints into e.g. the `PVForecastExternalStrategy` is not pretty, because we need lots of functionality of `ExternalMixin` in there. 
If you can not think of a better way, I propose to keep it like it is now and maybe deal with refactoring all classes and inheritance in a dedicated story. What do you think?

Reply:
I agree. I think that a wider refactoring would be good, even though it might need some discussion (it's quite a big topic).

Reply:
I agree that these classes desperately need refactoring (especially with the fact that these methods are not supported by all subclasses), however I think that this change is still compatible with the HomeMeter device external strategy. keep in mind that these are methods that can be overridden. Therefore all the HomeMeterExternalStrategy has to do in theory, is to override these methods and provide its own implementation. I promise to also try to refactor this in the context of D3ASIM-3549 (again), where I have started to make these 2 methods a bit more DRY.